### ISSUE TITLE
 FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN)

### ISSUE THREAD

**Issue by Independant7077** [Cisco-Talos/clamav#592]
Hello team,

we are consistently getting above error for most of our hosts for Clamav.
The error persist for few hours and gets resolved automatically  and again come back after some time and this is happening since past few days.

Though our clamav is up-to-date then also we are getting this error.

Please look into this issue as soon as possible!!

The error which we are getting is:

WARNING: FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
This means that you have been rate limited or blocked by the CDN.
 1. Verify that you're running a supported ClamAV version.
    See https://docs.clamav.net/faq/faq-eol.html for details.
 2. Run FreshClam no more than once an hour to check for updates.
    FreshClam should check DNS first to see if an update is needed.
 3. If you have more than 10 hosts on your network attempting to download,
    it is recommended that you set up a private mirror on your network using
    cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
    CDN and your own network.
 4. Please do not open a ticket asking for an exemption from the rate limit,
    it will not be granted.

Thanks,

**Comment by smaceno** [comment]
I've also been consistently receiving this message several times per day from several of our environments.  It looks to have started on or around 5/19/2022.  ClamAV 0.103.5.  I realize this message seems to be due to rate limiting, but we did not receive the error message prior to 5/19/2022, and had not made any changes to our environments at that time.

freshclam -vvv
Current working dir is /var/lib/clamav/
Loaded freshclam.dat:
  version:    1
  uuid:       7aa37448-e07f-469d-a2ab-bd90e798dfe2
  retry-after: 2022-05-24 13:42:05
ClamAV update process started at Tue May 24 09:44:14 2022
Current working dir is /var/lib/clamav/
Querying current.cvd.clamav.net
TTL: 1642
WARNING: DNS record is older than 3 hours.
WARNING: FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
This means that you have been rate limited or blocked by the CDN.
 1. Verify that you're running a supported ClamAV version.
    See https://docs.clamav.net/faq/faq-eol.html for details.
 2. Run FreshClam no more than once an hour to check for updates.
    FreshClam should check DNS first to see if an update is needed.
 3. If you have more than 10 hosts on your network attempting to download,
    it is recommended that you set up a private mirror on your network using
    cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
    CDN and your own network.
 4. Please do not open a ticket asking for an exemption from the rate limit,
    it will not be granted.
WARNING: You are still on cool-down until after: 2022-05-24 13:42:05

**Comment by roushanrjqikfox** [comment]
I am also getting same error while performing the very first database using freshclam. Please fix the issue.

**Comment by micahsnyder** [comment]
Hi @smaceno @roushanrjqikfox @Independant7077,

Some others have reported this issue as well in https://github.com/Cisco-Talos/clamav/issues/588

What appears to be new is this message:
```
Trying to retrieve CVD header from https://database.clamav.net/main.cvd
WARNING: remote_cvdhead: Malformed CVD header (too short)
WARNING: Failed to get main database version information from server: https://database.clamav.net
ERROR: check_for_new_database_version: Failed to find main database using server https://database.clamav.net.
```

Can you confirm if you are also seeing "remote_cvdhead: Malformed CVD header (too short)", before your freshclam was rate-limited?

I have a feeling the issue is bug in our Cloudflare CDN, because we haven't changed `main.cvd` is a long time. I am going to ask our Ops team to clear cache in our Cloudflare admin dashboard to see if that clears up issues for the users running into this. 

Regards,
Micah

**Comment by smaceno** [comment]
@micahsnyder Yes, here's an excerpt from one of my error messages if it helps, though I see the same errors, and more often for 'daily database'.

ERROR: check_for_new_database_version: Failed to find main database using server https://database.clamav.net/.
ERROR: check_for_new_database_version: Failed to find main database using server https://database.clamav.net/.
ERROR: remote_cvdhead: Malformed CVD header (too short)
ERROR: check_for_new_database_version: Failed to find main database using server https://database.clamav.net/.
ERROR: Update failed for database: main
ERROR: Database update process failed: HTTP GET failed
ERROR: Update failed.

"Daily database" error:

ERROR: check_for_new_database_version: Failed to find daily database using server https://database.clamav.net/.
ERROR: check_for_new_database_version: Failed to find daily database using server https://database.clamav.net/.
ERROR: remote_cvdhead: Malformed CVD header (too short)
ERROR: check_for_new_database_version: Failed to find daily database using server https://database.clamav.net/.
ERROR: Update failed for database: daily
ERROR: Database update process failed: HTTP GET failed
ERROR: Update failed.

**Comment by micahsnyder** [comment]
Hm so it's not just main, it's daily too.  I'm still working with our cloudflare admins to investigate.  Cloudflare themselves seem to think everything is o.k., other than some redirections: https://www.cloudflarestatus.com/

Can you please try using `freshclam --verbose` to get more detail about the failure?

**Comment by micahsnyder** [comment]
Another question - what OS are you using?  How did you install ClamAV?  It sounds like maybe DNS isn't being used to check the CVD version.

**Comment by smaceno** [comment]
@micahsnyder errors have been less frequent over the last couple of days.  Was worst between 5/19 & 5/25.  Actually has been pretty clean since then for the most part.  I'm using CentOS 7, and clam is installed via yum from the EPEL repository.  Configuration is standard.  We've been using it for a long time without incident until the date range noted.

**Comment by micahsnyder** [comment]
Thanks @smaceno  that helps. I just figured out the source of the issue.  If the clamav build fails to detect `libresolv`, it will build with DNS disabled.  Then every check for updated databases will count towards getting rate-limited.  

Issues in freshclam are two-fold:  
1. Now that we're rate-limiting HTTP requests, we really shouldn't allow the build to continue if `libresolv` can't be detected. 
2. The `remote_cvdhead()` function isn't smart enough to know it got a 429 rate-limiting response, so the error message is super misleading.

I suspect that one or more packages were recently built with DNS disabled, hence the influx of users running into this bug.  You said EPEL, which is very popular, which would really explain it.  I'll go test the EPEL package and see what's up.  Maybe I can reach out to EPEL maintainers to see if we can get it fixed in an update to the package. 

Then in some future clamav patch version, we can make it so libfreshclam can't be built without DNS support.

**Comment by micahsnyder** [comment]
My suspicion about DNS being disabled in the new packages was incorrect.  I confirmed that using DNS for the CVD version checks is working in the EPEL and ubuntu 0.103.6 packages.  I'm still unsure why a number of people saw it attempt to download the CVD header for the version check instead of using the DNS entry.  Perhaps a local DNS issue? 

Anyways... last week we also cleared cloudflare's cache of the `main.cvd` and `daily.cvd` files.  I haven't heard any new complaints since then on this issue.  

@smaceno, @Independant7077, @roushanrjqikfox are you still having issues or does it appear to be working again?

**Comment by JohanJNL** [comment]
I noticed last week that I didn't receive updates for a very long time. After I compiled the packages myself from the source package of Debian unstable, I could download again. But then I ran in a time-out. For some reason the download time-out in my freshclam.conf was set on 30 seconds, which is too short for my slow internet connection. With "ReceiveTimeout 60" I receive all updates again, including main.cvd.

**Comment by micahsnyder** [comment]
> I noticed last week that I didn't receive updates for a very long time. After I compiled the packages myself from the source package of Debian unstable, I could download again. But then I ran in a time-out. For some reason the download time-out in my freshclam.conf was set on 30 seconds, which is too short for my slow internet connection. With "ReceiveTimeout 60" I receive all updates again, including main.cvd.

Good to hear.  The `ReceiveTimeout 30` setting in the config is provided by some distribution packages.  If you remove the option, it will use the default which is `0` (disabled).

**Comment by JohanJNL** [comment]
Thanks, I will. Maybe compiling wasn't needed after all. It is possible that I got rate limited because my Freshclam kept trying to download main.cvd, but didn't succeed within 30 seconds.

**Comment by micahsnyder** [comment]
It looks to me like the event where cloudflare was serving up smaller files requested which was causing freshclam to retry and then get rate-limited has been resolved.  I'm going to close this issue.

If you observe this message again in advance of rate limiting:
`WARNING: remote_cvdhead: Malformed CVD header (too short)`
then feel free to re-open the issue.

**Comment by jessequinn** [comment]
Interesting, I have incorporated a mirror yet the error still appears

```sudo cat /etc/clamav/freshclam.conf
# Automatically created by the clamav-freshclam postinst
# Comments will get lost when you reconfigure the clamav-freshclam package
DatabaseOwner clamav
UpdateLogFile /var/log/clamav/freshclam.log
LogVerbose false
LogSyslog false
LogFacility LOG_LOCAL6
LogFileMaxSize 0
LogRotate true
LogTime true
Foreground false
Debug false
MaxAttempts 5
DatabaseDirectory /var/lib/clamav
DNSDatabaseInfo current.cvd.clamav.net
ConnectTimeout 30
ReceiveTimeout 0
TestDatabases yes
ScriptedUpdates yes
CompressLocalDatabase no
Bytecode true
NotifyClamd /etc/clamav/clamd.conf
# Check for new database 24 times a day
Checks 24
DatabaseMirror http://192.168.2.230:8080
DatabaseMirror http://192.168.2.230:8080
+ sudo sed -i 's/^DatabaseMirror .*$/DatabaseMirror http:\/\/192.168.2.230:8080/' /etc/clamav/freshclam.conf
+ sudo freshclam -v --config-file=/etc/clamav/freshclam.conf
Sat Jul 23 16:25:42 2022 -> *Current working dir is /var/lib/clamav/
Sat Jul 23 16:25:42 2022 -> *Loaded freshclam.dat:
Sat Jul 23 16:25:42 2022 -> *  version:    1
Sat Jul 23 16:25:42 2022 -> *  uuid:       35124cff-65be-4985-b34b-6c82c507acee
Sat Jul 23 16:25:42 2022 -> *  retry-after: 2022-07-24 00:00:36
Sat Jul 23 16:25:42 2022 -> ClamAV update process started at Sat Jul 23 16:25:42 2022
Sat Jul 23 16:25:42 2022 -> *Current working dir is /var/lib/clamav/
Sat Jul 23 16:25:42 2022 -> *Querying current.cvd.clamav.net
Sat Jul 23 16:25:42 2022 -> *TTL: 101
Sat Jul 23 16:25:42 2022 -> *fc_dns_query_update_info: Software version from DNS: 0.103.6
Sat Jul 23 16:25:42 2022 -> ^FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
Sat Jul 23 16:25:42 2022 -> This means that you have been rate limited or blocked by the CDN.
Sat Jul 23 16:25:42 2022 ->  1. Verify that you're running a supported ClamAV version.
Sat Jul 23 16:25:42 2022 ->     See https://docs.clamav.net/faq/faq-eol.html for details.
Sat Jul 23 16:25:42 2022 ->  2. Run FreshClam no more than once an hour to check for updates.
Sat Jul 23 16:25:42 2022 ->     FreshClam should check DNS first to see if an update is needed.
Sat Jul 23 16:25:42 2022 ->  3. If you have more than 10 hosts on your network attempting to download,
Sat Jul 23 16:25:42 2022 ->     it is recommended that you set up a private mirror on your network using
Sat Jul 23 16:25:42 2022 ->     cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
Sat Jul 23 16:25:42 2022 ->     CDN and your own network.
Sat Jul 23 16:25:42 2022 ->  4. Please do not open a ticket asking for an exemption from the rate limit,
Sat Jul 23 16:25:42 2022 ->     it will not be granted.
Sat Jul 23 16:25:42 2022 -> ^You are still on cool-down until after: 2022-07-24 00:00:36
+ sudo systemctl start clamav-freshclam
```

is `current.cvd.clamav.net` doing the rate limiting? How can we get around this as i have dozens of VMs behind nat so all appear from same IP.

**Comment by micahsnyder** [comment]
@jessequinn Freshclam's `freshclam.dat` file in the database directory records a timestamp to so prevent re-trying connections when freshclam knows it is being rate limited. The reason was we found that we were serving up a considerable amount of data in just 429 and 403 responses. So it is self-limiting for a time before it tries again.   

So if you just swapped over to the private mirror, you may need to delete the `freshclam.dat` file and try again.

**Comment by isaac-machakata** [comment]
My network is slow, cant download the [daily.cvd](https://database.clamav.net/daily.cvd) file before getting locked out
```
sudo freshclam
Tue Sep 13 23:13:34 2022 -> ClamAV update process started at Tue Sep 13 23:13:34 2022
Tue Sep 13 23:13:34 2022 -> ^Can't query current.cvd.clamav.net
Tue Sep 13 23:13:34 2022 -> ^Invalid DNS reply. Falling back to HTTP mode.
Tue Sep 13 23:13:34 2022 -> ^FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
Tue Sep 13 23:13:34 2022 -> This means that you have been rate limited or blocked by the CDN.
Tue Sep 13 23:13:34 2022 ->  1. Verify that you're running a supported ClamAV version.
Tue Sep 13 23:13:34 2022 ->     See https://docs.clamav.net/faq/faq-eol.html for details.
Tue Sep 13 23:13:34 2022 ->  2. Run FreshClam no more than once an hour to check for updates.
Tue Sep 13 23:13:34 2022 ->     FreshClam should check DNS first to see if an update is needed.
Tue Sep 13 23:13:34 2022 ->  3. If you have more than 10 hosts on your network attempting to download,
Tue Sep 13 23:13:34 2022 ->     it is recommended that you set up a private mirror on your network using
Tue Sep 13 23:13:34 2022 ->     cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
Tue Sep 13 23:13:34 2022 ->     CDN and your own network.
Tue Sep 13 23:13:34 2022 ->  4. Please do not open a ticket asking for an exemption from the rate limit,
Tue Sep 13 23:13:34 2022 ->     it will not be granted.
Tue Sep 13 23:13:34 2022 -> ^You are still on cool-down until after: 2022-09-14 16:52:32
```

**Comment by micahsnyder** [comment]
@isaac-machakata you can disable the "ReceiveTimeout" option in your `freshclam.conf` by setting
```
ReceiveTimeout 0
```

This should enable you to download the entire file.

**Comment by xorinzor** [comment]
@micahsnyder any chance the documentation can be updated with information about what Rate Limits are being enforced?
I got rate limited, but can't find any information whatsoever about what rules are being enforced. 

Even if it's just some basic generic information about X calls per hour, or day, or even a week would be useful since right now I don't know when I should bother to try again..

**Comment by tmorusu** [comment]
I did disable the timeout option in freshclam.conf however am still getting error.

# freshclam
ClamAV update process started at Fri Jun 16 06:24:58 2023
WARNING: FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
This means that you have been rate limited or blocked by the CDN.
 1. Verify that you're running a supported ClamAV version.
    See https://docs.clamav.net/faq/faq-eol.html for details.
 2. Run FreshClam no more than once an hour to check for updates.
    FreshClam should check DNS first to see if an update is needed.
 3. If you have more than 10 hosts on your network attempting to download,
    it is recommended that you set up a private mirror on your network using
    cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
    CDN and your own network.
 4. Please do not open a ticket asking for an exemption from the rate limit,
    it will not be granted.
WARNING: You are still on cool-down un

**Comment by besmirzanaj** [comment]
I wonder why this issue is closed. This is a persistent error.

**Comment by micahsnyder** [comment]
@besmirzanaj This issue is closed because each occurrence is unique to the reporter. 

If you're running into this issue and need help, feel free to open a new issue.  

What will help is if you can also include the output from `freshclam --verbose` after deleting the "freshclam.dat" file from your database directory (typically either `/var/lib/clamav/freshclam.dat` or `/usr/local/share/clamav/freshclam.dat`).  The verbose freshclam output should include details that we can use to triage the issue.

**Comment by Danila2577** [comment]
Hello! This happens when I try to run  $sudo freshclam. Is there any workaround?

```
Sun Aug  4 21:23:25 2024 -> ClamAV update process started at Sun Aug  4 21:23:25 2024
Sun Aug  4 21:23:25 2024 -> ^FreshClam previously received error code 429 or 403 from the ClamAV Content Delivery Network (CDN).
Sun Aug  4 21:23:25 2024 -> This means that you have been rate limited or blocked by the CDN.
Sun Aug  4 21:23:25 2024 ->  1. Verify that you're running a supported ClamAV version.
Sun Aug  4 21:23:25 2024 ->     See https://docs.clamav.net/faq/faq-eol.html for details.
Sun Aug  4 21:23:25 2024 ->  2. Run FreshClam no more than once an hour to check for updates.
Sun Aug  4 21:23:25 2024 ->     FreshClam should check DNS first to see if an update is needed.
Sun Aug  4 21:23:25 2024 ->  3. If you have more than 10 hosts on your network attempting to download,
Sun Aug  4 21:23:25 2024 ->     it is recommended that you set up a private mirror on your network using
Sun Aug  4 21:23:25 2024 ->     cvdupdate (https://pypi.org/project/cvdupdate/) to save bandwidth on the
Sun Aug  4 21:23:25 2024 ->     CDN and your own network.
Sun Aug  4 21:23:25 2024 ->  4. Please do not open a ticket asking for an exemption from the rate limit,
Sun Aug  4 21:23:25 2024 ->     it will not be granted.
Sun Aug  4 21:23:25 2024 -> ^You are still on cool-down until after: 2024-08-05 19:32:33
```

**Comment by micahsnyder** [comment]
> Hello! This happens when I try to run $sudo freshclam. Is there any workaround?

@Danila2577 please read the previous message and then make a new issue for support and provide the required details.  
Please also look at https://github.com/Cisco-Talos/clamav/issues/500 in case it applies to you.
